internal/dataplane/parser/parser.go

```diff
@@ -386,14 +390,134 @@ func getCertFromSecret(secret *corev1.Secret) (string, string, error) {
 	return cert, key, nil
 }
 
-func getCerts(log logrus.FieldLogger, s store.Storer, secretsToSNIs map[string][]string) []kongstate.Certificate {
-	snisAdded := make(map[string]bool)
-	// map of cert public key + private key to certificate
-	type certWrapper struct {
-		cert              kong.Certificate
-		CreationTimestamp metav1.Time
+type certWrapper struct {
+	identifier        string
+	cert              kong.Certificate
+	snis              []string
+	CreationTimestamp metav1.Time
+}
+
+func getGatewayCerts(log logrus.FieldLogger, s store.Storer) []certWrapper {
+	certs := []certWrapper{}
+	gateways, err := s.ListGateways()
+	if err != nil {
+		log.WithError(err).Error("failed to list Gateways")
+		return certs
```

Comment:
I don't think this should definitely change (there are reasons to prefer best effort and `parser.Build()` doesn't have final say on whether config can be applied; Kong can still reject it), but would argue that the scope here should be to make something consistent with the behavior of other parser components (which I think it does): if we want to refactor parser behavior more broadly, that should be a separate spike and discussion.

I agree that this is unlikely; it's a case of potentially being able to hit an error (here because client-go can potentially return one--it looks like only if the cached resource is not a proper Kubernetes object?) and needing to do something with it in a function that we don't want to return an error itself.

`parser.Build()` isn't consistent about using functions that log individual failures and always return successfully (possibly with an empty or partially full set), though that is the case for the majority (only `ingressRules.populateServices()` and `p.storer.ListCACerts()` errors will bubble up and break the build). `getCerts()` did skip individual certificates if it failed to load them, and `getCACerts()` just happens to have an additional conditions where it cannot return anything because it's populating from a top level resource (Gateways) rather than using outputs from another function (`getCerts()` receives a list of certificates from `ingressRulesFromIngressV1()`). The other top level resource queries do also return empty sets if their list fails.

IMO this best effort approach isn't unreasonable: if you have 99 good plugins (or consumers) and one that fails to load because it doesn't exist (a case where we currently log and omit it from the returned set), it does make sense to return those that are valid and proceed with those rather than breaking the build. This is particularly true in shared environments (common): for example, if the controller pulls from both `namespace-team-turtles` and `namespace-team-tortoises`, the turtles and tortoises teams can only access their respective namespaces, and the turtles helpfully typos their plugin name before collectively going on vacation, they'll block changes from either team and cause the tortoises considerable anguish and pain.

The cases that do break the build are also dubious: `ListCACerts()` could, like other top level resources, use a wrapper that returns an empty set (this will either break on apply because plugins reference unknown CA certs or fail safe at runtime because it cannot verify a client cert) and `populateServices()` [should only invalidate individual broken routes](https://github.com/Kong/kubernetes-ingress-controller/issues/2565). Invalid configuration de facto fails anyway down the road when Kong rejects it, and Kong is a better judge of that validity than we are.

Reply:
I appreciate your thoughtful response. There's definitely some truth to the argument that we shouldn't be blocking a bunch of valid configs for one problem, but I think the fact that the problem exists at all is sort of intrinsically part of a greater problem. At the end of the day however this PR didn't invent this wheel so it's not responsible for the wheel unless it broke it, which it didn't. So for now I'll consider this resolved as I don't intend it to be a blocker and I'll think on this some more for later.